### PR TITLE
Muh Nightvisuns! [Shady Zombies]

### PR DESCRIPTION
**Pull request by CIB**
- Added the NIGHT_INVISIBILITY flag. Monsters with this flag are not visible with only "indistinct" vision or lower (they basically use magic to meld into the shade).
- Added a new monster, shady zombie, who makes use of this flag. It is stat-wise equivalent to the normal zombie, and spawns about as regularly as the smoker zombie. "shady" is a reference to a youtuber I follow.
- Gameplay Effect: Raids at night become more dangerous, using flashlights over nightvision now has advantages, keeping your base well lit is now of greater importance.

Note that shady zombies are not made visible by night vision, and this is intentional. Mechanics-wise, they're supposed to make light sources vs night vision more of a pro/con kind of thing, instead of just always going for night vision. Lore-wise, the night invisibility actually turns a creature invisible, and the magic is "broken" by strong light.

### CODE REVIEW

**Comment by kevingranade** [comment]
Sounds good to me, eventually I'd like this to extend to a lot of monsters,
but this makes a good intro.

File: data/json/monsters.json [hunk 1]
```diff
@@ -2003,6 +2003,36 @@
     "flags":["SEES", "HEARS", "SMELLS", "STUMBLES", "WARM", "BASHES", "GROUP_BASH", "POISON", "HARDTOSHOOT", "NO_BREATHE", "REVIVES", "BONES", "PUSH_MON"]
     },{
     "type" : "MONSTER",
+    "id" : "mon_zombie_shady",
+    "name": "shady zombie",
+    "species":"ZOMBIE",
+    "default_faction":"zombie",
+    "symbol":"Z",
+    "color":"ltgray",
```
Comment (CIB) [inline data/json/monsters.json]: I don't use ASCII, would there be a better / unique color/symbol for it?

Comment (kevingranade) [inline data/json/monsters.json]: Unless it's a very unique zombie, stock with Z.
As for the colour, just not "the same color as shadows".

**Comment by illi-kun** [comment]
Please describe newly introduced flag 'NIGHT_INVISIBILITY' in JSON_FLAGS.md.

**Comment by illi-kun** [comment]
Hmm, and now I'm thinking about new unique monster - Jack of Shadows (obviously, that will be very hard enemy to beat)...

File: data/json/monsters.json (hunk 1 above)
Comment (CIB) [inline data/json/monsters.json]: Is ltgray the same color as shadows in ASCII? :) I just copied it from the smoker.. If nobody else has a suggestion I'll probably just pick a random color.

**Comment by CIB** [comment]
I've been thinking about a shadow-based faction (not zombies). They'd be breeding in dark places like basements and windowless houses (hello, hospitals!).

File: data/json/monsters.json [hunk 2]
```diff
@@ -2003,6 +2003,36 @@
...
+    "dodge":0,
+    "armor_bash":0,
+    "armor_cut":0,
+    "vision_day":3,
+    "vision_night":40,
+    "death_drops": "default_zombie_death_drops",
+    "hp":80,
+    "death_function":"NORMAL",
+    "special_attacks":[["BITE", 5], ["GRAB", 7]],
+    "description":"An uncanny shadow envelops this creature, as if light itself were too repulsed to touch it. All you can make out is its shambling, human shaped outline.",
```
Comment (illi-kun) [inline data/json/monsters.json]: Please use double spaces after each period.

File: src/mtype.h [hunk 3]
```diff
@@ -142,6 +142,7 @@ enum m_flag : int {
     MF_INTERIOR_AMMO,       // Monster contain's its ammo inside itself, no need to load on launch. Prevents ammo from being dropped on disable.
     MF_CLIMBS,              // Monsters that can climb certain terrain and furniture
     MF_PUSH_MON,            // Monsters that can push creatures out of their way
+    NIGHT_INVISIBILITY,     // Monsters that are invisible in poor light conditions
```
Comment (Coolthulhu) [inline src/mtype.h]: Monsterflags in this enum all have names starting with `MF_`.

**Comment by Skidborg** [comment]
Looks like an interesting concept. Probably good to expand it to a full zombie evolution chain.

**Comment by Nioca** [comment]
Mostly lurker here, popping in with a couple queries.

First off, how does this interact with Light Amp Goggles (and similar effects)? Would monsters with this flag show up under light amplification, or would they stay hidden?

Second, it strikes me as a bit odd to justify the abilities of Shady Zombies as magic, especially seeing as explicitly magical effects are kept at something of a minimum. Lore-wise, wouldn't it be better to just say that these zombies are better at blending in? Or are perhaps under the effects of something similar to the Artificial Night Generator CBM (albeit at a far weaker level)? Nether-style monsters could probably get away with it, but just regular zombies?

**Comment by kevingranade** [comment]
I'd prefer to say they're just camoflaged. I think what Christian means is
that there's no "stealth check" or similar, just a special ability.
I'm not sure which way to go with NV goggles from a thematic point of view,
but from a balance point of view making them totally invisible to NV
goggles is problematic.

**Comment by Skidborg** [comment]
Visible to night vision goggles makes sense.  Active camouflage may only work on the visible vision spectrum.   And there absolutely does need to be a way to deal with these things without turning the lights on and giving away your position. Night raiding could be more difficult, but stealth builds shouldn't be impossible.

**Comment by Rivet-the-Zombie** [comment]
>Night raiding could be more difficult, but stealth builds shouldn't be impossible.

A very good point.

File: data/json/monstergroups.json [hunk 4]
```diff
@@ -523,6 +523,8 @@
       { "monster" : "mon_zombie_hunter", "freq" : 10, "cost_multiplier" : 5 },
       { "monster" : "mon_skeleton", "freq" : 30, "cost_multiplier" : 5 },
       { "monster" : "mon_zombie_smoker", "freq" : 10, "cost_multiplier" : 5 },
+      { "monster" : "mon_zombie_shady", "freq" : 10, "cost_multiplier" : 4},
+      { "monster" : "mon_zombie_shady", "freq" : 30, "cost_multiplier" : 15, "pack_size": [4,6]},
```
Comment (RadaRadaRada) [inline data/json/monstergroups.json]: This says that, groups of shadies are more common then lone ones. Is this intentional?

File: data/json/monsters.json (hunk 1 above)
Comment (Coolthulhu) [inline data/json/monsters.json]: White background could work well.
On black background, it would need a bright foreground instead.

**Comment by CIB** [comment]
> but from a balance point of view making them totally invisible to NV goggles is problematic.

Hm.. Maybe they could produce some kind of field every now and then that gives away their presence? It'd still make it hard to pinpoint their numbers, but at least you'd know they're around.

**Comment by Coolthulhu** [comment]
No, fields are a bad idea here.
Consider player performing a long action in darkness: a field appearing somewhere won't trigger an interrupt, even if the same field would easily be spotted by an active player.
Also, fields aren't visible on radar, meaning that low resolution would be a disadvantage.

Either make it about sounds (shriek attack would fit here well) or have invisibility randomly fail.

Adding shriek wouldn't unbalance them, because their stats are so low they're pretty much bound to be support tier. They could be an evolution of shrieker, who has a similar role - nighttime support.

Visual cues would be rather hard to balance. If the frequency of them was too low, they could easily approach the player without triggering them once. If it was too high, their invisibility would protect them only from ranged attacks.

That 40 night vision is quite a lot. It will probably work against them - they will be easy to separate from the horde and when alone, their stats will make them nearly harmless.

**Comment by CIB** [comment]
> Consider player performing a long action in darkness: a field appearing somewhere won't trigger an interrupt, even if the same field would easily be spotted by an active player.

Yes, I'm considering it, and I'm thinking to myself "That seems a pretty stupid thing to do with zombies around". I get that night vision is supposed to be good for stealth raids, but for performing long term activities without setting up some kind of lighting to alert you of enemies, not so much.

**Comment by Coolthulhu** [comment]
It is still a bad idea to use something as "missable" as fields. It offers no advantage over failing invisibility, but would certainly be more annoying.
Long actions start with actions as short as reloading a pistol or applying first aid. There will be more later on, as simple actions become long actions for convenience (for example, using handle to open a garage door).

**Comment by chaosvolt** [comment]
Hmm. This could potentially be added to shadows and the wraith. X3

**Comment by kevingranade** [comment]
The problem worn NV goggles I was anticipating was that they totally
override normal vision, so even adjacent shadow zombies would remain
invisible.
Now that I think about it though, they should be making enough noise to be
detected at a few squares distance, so it's less of a problem than I
thought.
If you're spending a lot of time in the dark and you can't hear, you were
asking door it :)

**Comment by chaosvolt** [comment]
Huehuehue.

I would hope thermal vision can foil them though?

**Comment by Skidborg** [comment]
A thought: Could the detection of NIGHT_INVISIBILITY creatures scale with perception? It seems like the kind of thing the stat should effect, and make it a less easy stat to dump.

**Comment by Coolthulhu** [comment]
> A thought: Could the detection of NIGHT_INVISIBILITY creatures scale with perception?

That would probably require a more complete system of limited visibility.
With the current implementation, combining monster vision and player vision with limited visibility would probably be quite clunky and bug prone.

**Comment by sparr** [comment]
Loving the effect this concept has on dark areas, night vision, etc.

**Comment by CIB** [comment]
:| I had meant to close this.. D: Oops! I don't think it's necessarily ready with regards to night vision..

**Comment by Coolthulhu** [comment]
It's a bit tougher than expected on weak characters now because of the whole "always totally invisible" thing, but it may still be OK.

I just realized that there's one big problem with sounds and this: the game won't generate footstep "signals" if the source tile is seen. Meaning that no-NV characters will get a warning, but the NV ones will not hear anything.

**Comment by chaosvolt** [comment]
Hmm. It might be sensible to give the monster a parrot feature so you occasionally hear murmuring or other "spooky" sounds to give this creature away. :V

**Comment by kevingranade** [comment]
Minor correction, players will still be able to hear monsters, but it won't
draw an indicator for its position.
That's still a significant problem, suggestions for fixing it are welcome.

**Comment by chaosvolt** [comment]
Ah. Right, that makes sense. Sorry. x.x

Most likely would require some addition of a "is source a monster with the invisi-flag" check to the "check if footprint source is in sight" function.

**Comment by Skidborg** [comment]
Doing some preliminary testing with a weak, stealth-based character and they don't seem to be too hard at the moment.  They add some scare value when they suddenly appear out of nowhere, but as long as they don't manage to get a grab on that first turn of contact it's trivially easy to just slip away.  

If they were more powerful and were loud enough to leave an intermittent sound trail at long range it would increase the tension quite a bit.

Honestly, the biggest thing I don't like about them is the name.  Not very descriptive or menacing.

**Pull request merged**